simdev/module/github/github_api_wrapper.py

```diff
@@ -0,0 +1,186 @@
+        :param progress: TQDM progress if any
+        :return: set of stargazers
+        """
+        if progress is not None:
+            progress.desc = f"Fetching stargazers for {repo}"
+
+        fixed_url_part = (
+            f"{self._api_url}repos/{repo}/stargazers?per_page={str(batch_count)}&page="
+        )
+        current_page = 0
+        result: Set[str] = set()
+
+        while True:
+            if progress is not None:
+                progress.set_postfix_str("page " + str(current_page + 1))
+            page_response: List[dict] = self._get_request_json(
+                fixed_url_part + str(current_page)
+            )
+            if len(page_response) == 0:
+                break
+            if not isinstance(page_response, list):
+                logging.warning(
+                    "Failure fetching stargazers for %s:"
+                    "\nRequest: %s"
+                    '\nUnknown response: "%s"',
+                    repo,
+                    fixed_url_part + str(current_page),
+                    json.dumps(page_response),
+                )
+                continue
+            for stargazer in page_response:
+                result.add(stargazer["login"])
+            current_page += 1
+            if current_page > page_limit:
+                break
+        return result
+
+    def fetch_starred_repositories(
+        self, user, batch_count=100, page_limit=400, progress=None
+    ) -> Set[str]:
+        """
+        Fetch repositories starred by a user
+        :param user: to fetch starred repositories of
+        :param batch_count: number of entries per GitHub API response's page
+        :param page_limit: max amount of pages to fetch
+        :param progress: TQDM progress
+        :return: set of starred repositories
+        """
+        if progress is not None:
+            progress.desc = f"Fetching starred repositories of {user}"
+
+        fixed_url_part = (
+            f"{self._api_url}users/{user}/starred?per_page={batch_count}&page="
+        )
+        current_page = 0
+        result = set()
+
+        while True:
+            page_response: starred_response_type = None
+            try:
+                if progress is not None:
+                    progress.set_postfix_str("page " + str(current_page + 1))
+                page_response = self._get_request_json(
+                    fixed_url_part + str(current_page)
+                )
+                if len(page_response) == 0 or (
+                    "message" in page_response
+                    and page_response["message"] == "Not Found"
+                ):
+                    break
+                for repo in page_response:
+                    result.add(repo["full_name"])
+            except Exception:
+                if (
+                    page_response is not None
+                    and "message" in page_response
+                    and "rate limit" in page_response["message"]
```

Comment:
I would recommend to move logic like this to separate functions like `is_rate_limit_reached` - it will make easier to understand code + it's much easier to test